text/0002-serializable-types.md

```diff
@@ -0,0 +1,413 @@
+void WorldClientRpc(MyComplexType[] values) { /* ... */ }
+```
+
+### NetworkObject & NetworkBehaviour
+
+`NetworkObject` and `NetworkBehaviour` instances will be serialized by built-in serialization code if instances are not `null` and `NetworkObject.IsSpawned == true`. Passing a `null` instance or a `NetworkObject` with `IsSpawned == false` or a `NetworkBehaviour` with `NetworkObject.IsSpawned == false` will still lead into RPC replication but remote side will have `null` values for the parameters. IDs of spawned `NetworkObject` and `NetworkBehaviour` instances will be resolved by running `NetworkManager` instance, and those IDs will be the links between local and remote instances. Also, those IDs will be used when serializing `NetworkObject` and `NetworkBehaviour` instances as a part of an RPC call.
+
+```cs
+[ServerRpc]
+void MyServerRpc(NetworkObject nobj, NetworkBehaviour nbhv) { /* ... */ }
+
+void Update()
+{
+    if (Input.GetKeyDown(KeyCode.P))
+    {
+        MyServerRpc(/* NetworkObject = */ this.NetworkObject, /* NetworkBehaviour = */ this); // Client -> Server
+    }
+}
+```
+
+### INetworkSerializable & BitSerializer
+
+Complex user-defined types that implements `INetworkSerializable` interface will be serialized by user provided serialization code.
+
+An instance of `BitSerializer` will be passed into `INetworkSerializable::NetworkSerialize(BitSerializer)` method which can be used to easily serialize fields by reference.
+
+All types supporting serialization will also be supported by `BitSerializer` with `BitSerializer::Serialize(ref int value)` variant methods and templated `BitSerializer::Serialize<T>(ref T value) where T : INetworkSerializable` method.
+
+```cs
+class BitSerializer
+{
+    bool IsReading { get; }
+
+    void Serialize(ref int value) { /* ... */ }
+    void Serialize(ref float value) { /* ... */ }
+    // ...
+    void Serialize<T>(ref T value) where T : INetworkSerializable { /* ... */ }
+}
+
+interface INetworkSerializable
+{
+    void NetworkSerialize(BitSerializer serializer);
+}
+
+struct MyComplexStruct : INetworkSerializable
+{
+    public Vector3 Position;
+    public Quaternion Rotation;
+
+    // INetworkSerializable
+    public NetworkSerialize(BitSerializer serializer)
+    {
+        serializer.Serialize(ref Position);
+        serializer.Serialize(ref Rotation);
+    }
+    // ~INetworkSerializable
+}
+
+[ServerRpc]
+void MyServerRpc(MyComplexStruct myStruct) { /* ... */ }
+
+void Update()
+{
+    if (Input.GetKeyDown(KeyCode.P))
+    {
+        MyServerRpc(
+            new MyComplexStruct
+            {
+                Position = transform.position,
+                Rotation = transform.rotation
+            }); // Client -> Server
+    }
+}
+```
+
+#### Conditional Serialization
+
+As the developer has more control over serialization of a struct, one might implement conditional serialization at runtime.
+
+We will explore more advanced use-cases with the examples below:
+
+##### Example: Array
```

Comment:
I guess I wonder, could we have a built-in array?  if the array is just a sequence of INetworkSerializable types, then could we give users out-of-the-box array support vs. having to write an array handler?

Reply:
+1, this seems like the type of code that could be generalized in BitSerializer. Although it adds another layer of serialization code needed for each type supported which can't really be encapsulated

Reply:
we will provide static array serialization out-of-the-box as outlines above in this RFC, this array example is just serving a demonstration purpose here.